src/coreComponents/unitTests/fluidFlowTests/testCompFlowUtils.hpp

```diff
@@ -212,6 +219,574 @@ void setupProblemFromXML( ProblemManager & problemManager, char const * const xm
+
+      string const & fluidName = subRegion.getReference< string >( CompositionalMultiphaseBase::viewKeyStruct::fluidNamesString() );
+      MultiFluidBase const & fluid = subRegion.getConstitutiveModel< MultiFluidBase >( fluidName );
+      arrayView1d< string const > const & components = fluid.componentNames();
+
+      arrayView2d< real64, compflow::USD_COMP > const compDens =
+        subRegion.getExtrinsicData< extrinsicMeshData::flow::globalCompDensity >();
+
+      arrayView2d< real64, compflow::USD_COMP > const dCompDens =
+        subRegion.getExtrinsicData< extrinsicMeshData::flow::deltaGlobalCompDensity >();
+
+      arrayView2d< real64, compflow::USD_COMP > const compFrac =
+        subRegion.getExtrinsicData< extrinsicMeshData::flow::globalCompFraction >();
+
+      arrayView3d< real64, compflow::USD_COMP_DC > const dCompFrac_dCompDens =
+        subRegion.getExtrinsicData< extrinsicMeshData::flow::dGlobalCompFraction_dGlobalCompDensity >();
+
+      // reset the solver state to zero out variable updates
+      solver.resetStateToBeginningOfStep( domain );
+
+      // make a copy of unperturbed values of component fractions
+      array2d< real64, compflow::LAYOUT_COMP > compFracOrig( subRegion.size(), numComp );
+      compFracOrig.setValues< serialPolicy >( compFrac );
+
+      // update component density and check derivatives
+      for( integer jc = 0; jc < numComp; ++jc )
+      {
+        // reset the solver state to zero out variable updates (resetting the whole domain is overkill...)
+        solver.resetStateToBeginningOfStep( domain );
+
+        // perturb a single component density in each cell
+        forAll< serialPolicy >( subRegion.size(), [=] ( localIndex const ei )
+        {
+          real64 const dRho = perturbParameter * ( compDens[ei][jc] + perturbParameter );
+          dCompDens[ei][jc] = dRho;
+        } );
+
+        // recompute component fractions
+        solver.updateComponentFraction( subRegion );
+
+        // check values in each cell
+        forAll< serialPolicy >( subRegion.size(), [=] ( localIndex const ei )
+        {
+          SCOPED_TRACE( "Element " + std::to_string( ei ) );
+
+          auto dZ_dRho = invertLayout( dCompFrac_dCompDens[ei].toSliceConst(), numComp, numComp );
+          string var = "compDens[" + components[jc] + "]";
+
+          checkDerivative( compFrac[ei].toSliceConst(),
+                           compFracOrig[ei].toSliceConst(),
+                           dZ_dRho[jc].toSliceConst(),
+                           dCompDens[ei][jc],
+                           relTol,
+                           "compFrac",
+                           var,
+                           components );
+        } );
+      }
+    } );
+  } );
+}
+
+void testPhaseVolumeFractionNumericalDerivatives( CompositionalMultiphaseFVM & solver,
+                                                  DomainPartition & domain,
+                                                  bool const isThermal,
```

Comment:
So the `isThermal` is currently never `true`?

Reply:
Yes, I confirm that the `isThermal` is currently never `true` (until PR #1225 is merged). I am trying to take small steps towards thermal to avoid doing everything in the same PR. So currently,
- There is already a flag `isThermal` in the solver, currently always `false`.
- There is already some extrinsic data available in the solver for temperature, currently only modified at initialization.
- The present PR adds the unit tests functions to check derivatives wrt temperature, but they are currently unused.

PR #1225 will put everything together, allow the flag `isThermal` to be `true`, allow the temperature field to change with time, and will add a unit test using the functions added in the present PR.